### PR TITLE
stm32f4: SDIO support

### PR DESCRIPTION
**Pull request by CyberDeck**
Adds sdio support for the stm32f4 to allow for SD card flash updates without power cycling some boards, e.g. BTT Octopus Pro.

### Reason
You are not able to use the `spi_flash.py` script to update some boards like the BTT Octopus Pro because the BTT bootloader accesses the SD card in SDIO mode and not in SPI mode.
If you update the firmware by running the normal `spi_flash.py` script, it puts the SD card into SPI mode.
Then the board is reset but the bootloader tries to access the SD card in SDIO mode. However, the SD card itself is still in SPI mode and it cannot be switched back to SDIO mode without power cycling the SD card.
Please note that this behavior is correct according to the respective SD card standard. There is no way back to SDIO mode from SPI mode.

This patch introduces SDIO capabilities to klipper and provides an enhanced `spi_flash.py` to be able to speak in SDIO mode to SD cards.  

Having SDIO support was mentioned in the past, for exampe by @Arksine in #4530.

### Result
You are able to easily use `spi_flash.py` to update the firmware without the need for a power cycle.

Please note that this PR does not only introduce SDIO support in the stm32f4 firmware but also patches the existing `spi_flash.py` to have capabilities to flash by SPI and SDIO SD card commands.

### Example of a firmware update session
```bash
$ sudo service klipper stop
$ scripts/flash-sdcard.sh /dev/serial/by-id/usb-Klipper_stm32f446xx_40004A000451373330333137-if00 btt-octopus-pro
Flashing /home/pi/klipper-test/out/klipper.bin to /dev/serial/by-id/usb-Klipper_stm32f446xx_40004A000451373330333137-if00
Checking FatFS CFFI Build...
Connecting to MCU...Connected
Checking Current MCU Configuration...Done
MCU needs restart: is_config=1, is_shutdown=0
Attempting MCU Reset...Done
Waiting for device to reconnect.....Done
Connecting to MCU...Connected
Initializing SD Card and Mounting file system...

SD Card Information:
Version: 2.0
SDHC/SDXC: True
Write Protected: False
Sectors: 31116288
manufacturer_id: 3
oem_id: SD
product_name: SC16G
product_revision: 8.128
serial_number: 337605E5
manufacturing_date: 1/2018
capacity: 14.8 GiB
fs_type: FAT32
volume_label: BTTOCTOPRO
volume_serial: 851174759
Uploading Klipper Firmware to SD Card...Done
Validating Upload...Done
Firmware Upload Complete: firmware.bin, Size: 28400, Checksum (SHA1): C8F9BF096F7AB0A7E46A46B34F06FCAEAB8E389B
Attempting MCU Reset...Done
Waiting for device to reconnect.....Done
Connecting to MCU...Connected
Verifying Flash...Version matched...Done
Firmware Flash Successful
Current Firmware: v0.10.0-577-g1092de3c
Attempting MCU Reset...Done
SD Card Flash Complete
$ sudo service klipper start
```

### CODE REVIEW

**Comment by Arksine** [comment]
I think this is going to come down to whether or not Kevin wants to maintain low level SDIO support in Klipper that is only functional for this specific use case.   If he's interested in proceeding I'll do a full review.  I'm familiar as I started experimental SDIO support in CanBoot.

**Comment by Dviros** [comment]
Thanks for this @CyberDeck, what should be the steps required to make it work for STM32H743?
I'm not too sure about the .c syntax involved to be honest :\

**Comment by KevinOConnor** [comment]
> I think this is going to come down to whether or not Kevin wants to maintain low level SDIO support in Klipper that is only functional for this specific use case.

In general, I don't see an issue supporting SDIO in the mcu code.

-Kevin

**Comment by CyberDeck** [comment]
> what should be the steps required to make it work for STM32H743?

That is a really good question, @Dviros. The STM32 Datasheets are really tedious to read with around ~3300 pages.

In an early phase, I developed the patch on a STM32F4x development board. It was quite handy to have a in-circuit debugger at hand.
However, if the assumption is right, that the STM32F4x and the STM32H743 share the same IP on the chip for SD/SPI/SDIO stuff, the patch could work out of the box. I looked through the datasheets and no difference jumped into my eyes.

Are you targeting the SKR 3 EZ board with the STM32H743? I recently modified an old printer with one of this boards and I could test it in the next days. Originally, this patch was meant for my Voron 2.4 printer. I hated it to "turn it on it's back" to access the SD card for updates. Just drop me a note if you are interested in a SKR EZ 3 board test.

**Comment by Dviros** [comment]
I guess you guessed it right! Well done. Yes, I'm indeed asking about the BTT SKR 3.
At the moment, for the rare cases of actually upgrading the board, I have it take the printer's lower case off and connect to it via DFU, as I don't have an easy access to the SD card (and I suspect it does not work), so I don't want you to work too hard just for me.
Also, not to complain or something, I'm not familiar with another remote flashing technique that does not require physically mounting a card.
For example (and that may have been discussed many times; forgive me if I'm stating the obvious) Marlin has M997 to get boards in DFU.
Is there a way to emulate the BOOT button press or just reboot to DFU remotely using Klipper?
Again, thanks a lot for this.
D

**Comment by CyberDeck** [comment]
> Yes, I'm indeed asking about the BTT SKR 3.

I tried it with an BTT SKR 3 EZ board which uses the same STM32 chip (STM32H743xx).
My assumption so far was wrong. The STM32H7 series does not share the same IP for SD/SPI/SDIO stuff with the STM32F4 series. The STM32H7 has a dedicated SDMMC IP core which differs from the SDIO IP core the STM32F4 uses...

However, in a two hours quick hack session I could add compatibility for this IP core, too.

```bash
pi@easyframe:~/klipper-dev $ ./scripts/flash-sdcard.sh /dev/serial/by-id/usb-Klipper_stm32h743xx_4B0031000951303230373534-if00 btt-skr-3
Flashing /home/pi/klipper-dev/out/klipper.bin to /dev/serial/by-id/usb-Klipper_stm32h743xx_4B0031000951303230373534-if00
Checking FatFS CFFI Build...
Connecting to MCU...Connected
Checking Current MCU Configuration...Done
Initializing SD Card and Mounting file system...

SD Card Information:
Version: 2.0
SDHC/SDXC: False
Write Protected: False
Sectors: 245760
manufacturer_id: 0
oem_id: 0
product_name: APPSD
product_revision: 0.0
serial_number: 0000008B
manufacturing_date: 5/2019
capacity: 120.0 MiB
fs_type: FAT16
volume_label: Test
volume_serial: 1266771126
Uploading Klipper Firmware to SD Card...Done
Validating Upload...Done
Firmware Upload Complete: firmware.bin, Size: 27560, Checksum (SHA1): D89C5E8BF92CBC96DCB554F7E4A0EE7EEB5CA285
Attempting MCU Reset...Done
Waiting for device to reconnect......Done
Connecting to MCU...Connected
Verifying Flash...Version matched...Done
Firmware Flash Successful
Current Firmware: v0.10.0-575-gefa336f4
Attempting MCU Reset...Done
SD Card Flash Complete
```
Right now, it is only integrated within my [klipper playground](/CyberDeck/klipper/tree/sdmmc). You can try my `sdmmc` branch at your own risk.
Give me some time to merge it into this PR.

**Comment by Arksine** [comment]
I'm currently in the process of reviewing this PR, I should have it done either tonight or tomorrow.  I would recommend holding off on adding H7 support, it would probably be better to submit it as a separate pull request after this is merged.  I am going to recommend some changes and there are some things I want to ask Kevin for feedback on.

FWIW, there are merge conflicts that need to be resolved as well.

**Comment by Arksine** [comment]
I have added some initial review comments.  Overall the sdio implementation looks similar to what I came up, however I did want to bring up that when researching the stm32 SDIO peripheral it seems recommended to check half empty flags when reading from and writing to the FIFO.  Apparently this reduces the chance of buffer underrun/overrun.

File: src/stm32/gpio.h
Comment (KevinOConnor) [inline src/stm32/gpio.h:76]: I'd add these to a new "sdio.h" file.  I don't think we should modify "gpio.h" for this.

File: src/stm32/gpioperiph.c
Comment (KevinOConnor) [inline src/stm32/gpioperiph.c:46]: If it is necessary to configure the ospeed parameter, then add a new `GPIO_HIGHSPEED` flag to the existing `mode` parameter.  Also, please don't claim a copyright on gpio.c for such a minor change.

**Comment by CyberDeck** [comment]
Thank you for all your valuable input on my PR. I will change my PR according to your helpful review comments. Right now I am quiet often away from my printer not being able to test the changes. Please be patient...

**Comment by Arksine** [comment]
I went ahead and reopened since Cyberdeck indicated that it may take a bit of time to update the PR.

File: src/stm32/stm32f4.c
Comment (CyberDeck) [reply src/stm32/stm32f4.c:164]: Proposed changes included in latest commit.

File: scripts/spi_flash/spi_flash.py
Comment (CyberDeck) [reply scripts/spi_flash/spi_flash.py]: Proposed changes included in latest commit.

File: src/sdiocmds.c
Comment (CyberDeck) [reply src/sdiocmds.c]: Proposed changes included in latest commit.

File: src/sdiocmds.c
Comment (CyberDeck) [reply src/sdiocmds.c:45]: Proposed changes included in latest commit.

File: src/sdiocmds.c
Comment (CyberDeck) [reply src/sdiocmds.c:80]: Proposed changes included in latest commit.

File: src/stm32/sdio.c
Comment (CyberDeck) [reply src/stm32/sdio.c]: Proposed changes included in latest commit. Current speed included in the sdiodev_s struct and timeout calculation is done at command_sdio_read_data and command_sdio_write_data.

File: src/sdiocmds.c
Comment (CyberDeck) [reply src/sdiocmds.c:110]: Proposed changes included in latest commit.

File: src/stm32/sdio.c
Comment (CyberDeck) [reply src/stm32/sdio.c]: Removed double line.

File: src/stm32/gpio.h
Comment (CyberDeck) [reply src/stm32/gpio.h]: Proposed changes included in latest commit.

File: src/stm32/gpioperiph.c
Comment (CyberDeck) [reply src/stm32/gpioperiph.c]: Proposed changes included in latest commit. Removed the copyright claim. Sorry for introducing it for such a small change. I was blind there when I added the copyright claims to my changed / introduced files.

**Comment by CyberDeck** [comment]
> I have added some initial review comments. Overall the sdio implementation looks similar to what I came up, however I did want to bring up that when researching the stm32 SDIO peripheral it seems recommended to check half empty flags when reading from and writing to the FIFO. Apparently this reduces the chance of buffer underrun/overrun.

I had no luck for implementing it with the FIFO buffer half full / half empty flags. Even when I considered that one has to read / write at each time 8 words to the FIFO buffer I had no success. I think I am missing a bit there but the current implementation will work reliable in my opinion. But this is space for further improvements.

**Comment by CyberDeck** [comment]
Latest commit include a merge of the current upstream master and the proposed reviewer changes except that FIFO buffer stuff. I tested it on my BTT Octopus Pro board with a STM32F446 chip. Therefore I updated the boards_defs.py accordingly. It should work for the other btt-octopus derivates as well. However, I do not own these boards and someone probably should check it before including it within the boards_defs.py as well.

As soon as this will get merged into master, I will rework the SDMMC stuff into it as mentioned above for the STM32H743xx chip (e.g. BTT SKR 3 EZ boards).

**Comment by CyberDeck** [comment]
> I went ahead and reopened since Cyberdeck indicated that it may take a bit of time to update the PR.

@Arksine: Any opinion about the adapted PR?

**Comment by KevinOConnor** [comment]
It looks like this PR will need to wait until after the v0.11.0 release.  Otherwise, I'm open to merging once the review is complete.

-Kevin

**Comment by KevinOConnor** [comment]
What is the status of this PR?

-Kevin

**Comment by CyberDeck** [comment]
The last commit [f1af443](https://github.com/Klipper3d/klipper/pull/5759/commits/f1af443c09ac7d21398acc37a3df6e9c08a42d93) from end of october adressed all initial review comments so far. In my opinion, it is up to you and arksine to decide if they are addressed successfully.

**Comment by Darqeternal1** [comment]
Not to side track the conversation too much but does this change apply to the stm32f429?  In the files changed only the stm32f446 is modified.

**Comment by CyberDeck** [comment]
>... but does this change apply to the stm32f429?

It should be compatible with the stm32f429 as well. The official migration guide from stm32f429 to stm32f446 line does not show any big road blockers as long as you clock the sdio with the PLL48CLK, which I do.

But as I do not own such a board to test, I really do like to have the stm32f446 patch included first, to have a clean starting point for further patches like stm32f429 and stm32h743x. Then a volunteer could test those patches before including it to the main source tree.

**Comment by joaobarros** [comment]
> > ... but does this change apply to the stm32f429?
> 
> It should be compatible with the stm32f429 as well. The official migration guide from stm32f429 to stm32f446 line does not show any big road blockers as long as you clock the sdio with the PLL48CLK, which I do.
> 
> But as I do not own such a board to test, I really do like to have the stm32f446 patch included first, to have a clean starting point for further patches like stm32f429 and stm32h743x. Then a volunteer could test those patches before including it to the main source tree.

I have boards with all 3 chips you mentioned so feel free to ping me when testing stage is reached.

**Comment by Arksine** [comment]
Sorry for the delay.  After a bit of a hiatus I'm currently going through a load of notifications and noticed that this PR is pending my review.  I will attempt to complete it soon, next week at the latest.

**Comment by r0bb10** [comment]
> > ... but does this change apply to the stm32f429?
> 
> It should be compatible with the stm32f429 as well. The official migration guide from stm32f429 to stm32f446 line does not show any big road blockers as long as you clock the sdio with the PLL48CLK, which I do.
> 
> But as I do not own such a board to test, I really do like to have the stm32f446 patch included first, to have a clean starting point for further patches like stm32f429 and stm32h743x. Then a volunteer could test those patches before including it to the main source tree.

gonna test on the F429 also to see if it works, subscribing for now.

**Comment by KevinOConnor** [comment]
FWIW, this PR looks okay to me.  I only gave it a superficial review though.  Hopefully @arksine can provide any additional feedback - in particular on the changes to spi_flash.py .

-Kevin

**Comment by Arksine** [comment]
The changes look ok to me.  Quite a few new MCU commands and responses are introduced, if need be I think there may be room to optimize this in the future, however that is something we can look it if it becomes necessary.  Likewise we can see what happens in practice, if no one experiences buffer errors then there is no need to implement this using the half full and half empty flags.

**Review by Arksine — APPROVED** [review]

**Comment by KevinOConnor** [comment]
Thanks.  Looks like this PR just needs a rebase (to fix some minor merge issues) and confirmation from @CyberDeck that it is still good to be merged.

-Kevin

**Comment by r0bb10** [comment]
@CyberDeck any news on this PR?

**Comment by CyberDeck** [comment]
I updated the PR to the latest upstream master.
So far I tested it on my machine without any problems.

If you try to flash the firmware for the first time you should follow a specific sequence:
1. Update klipper sources on your machine and make the new firmware.
2. Manually edit the `scripts/spi_flash/board_defs.py`  and change the definition back to "swspi".
```    'btt-octopus-f446-v1': {
        'mcu': "stm32f446xx",
        'spi_bus': "swspi",
        'spi_pins': "PC8,PD2,PC12",
        'cs_pin': "PC11",
        'skip_verify': True
    },
```
3. Flash the file with the usual `scripts/flash-sdcard.sh /dev/serial/by-id/usb-Klipper_stm32f446x-xyc-if00 btt-octopus-pro-f446-v1.0`
4. Power off and power on the machine. (If this is not done the sdcard is still in spi mode...)
5. Change back the `scripts/spi_flash/board_defs.py` to the original entry:
```    'btt-octopus-f446-v1': {
        'mcu': "stm32f446xx",
        'sdio_bus': 'sdio'
    },
```
6. Try to reflash it now with SDIO. Everything should work.

Alternatively, one can update the firmware manually by putting it onto the SD card to have not switch back to SWSPI mode once. Following updates will not need this sequence. Only once to the the firmware with the SDIO support on the board.

Not sure if that was clear or helpful. If you run into probs just drop me a note.

**Comment by KevinOConnor** [comment]
Thanks.  I committed this change (commit d7bd7f1f).  However, I did not commit the change to board_defs.py as that looks to me like a non-backwards compatible change that would need an update to Config_Changes.md to document it.  (It's fine to make that change - it just needs the documentation to go along with it.)

-Kevin

**Comment by Pylonper** [comment]
Thanks all for all great work you are doing here!
I tried this yesterday om my voron with octopus 1.1 board. I kind of got it working by modifying the board_defs.py to use sdio, but the script hangs at the point when it uploads the firmware... 
However, if I let it be for a while and then exit the script with ctl c, and then run the script again with option -c the scripts confirms that the upload actually went fine.
Please let me know if I can help with more information
BR /Per

**Comment by Pylonper** [comment]
> Thanks. I committed this change (commit [d7bd7f1](https://github.com/Klipper3d/klipper/commit/d7bd7f1f4ba6cecd19daa566fdc1864561269ae1)). However, I did not commit the change to board_defs.py as that looks to me like a non-backwards compatible change that would need an update to Config_Changes.md to document it. (It's fine to make that change - it just needs the documentation to go along with it.)
> 
> -Kevin
@CyberDeck, did you see this? (not sure how github works now when pull request is closed) As def file is not committed, it won't work for general users

**Pull request closed without merging**
